src/szk.rs

```diff
@@ -0,0 +1,574 @@
+        let leader_proof_share = SzkProofShare::Leader {
+            uncompressed_proof_share: leader_proof_share,
+            leader_blind_and_helper_joint_rand_part,
+        };
+        let helper_proof_share = SzkProofShare::Helper {
+            proof_share_seed_and_blind: helper_seed.clone(),
+            leader_joint_rand_part,
+        };
+        Ok([leader_proof_share, helper_proof_share])
+    }
+
+    fn query(
+        &self,
+        input_share: &[T::Field],
+        proof_share: SzkProofShare<T::Field, SEED_SIZE>,
+        verify_key: &[u8; SEED_SIZE],
+        nonce: &[u8; 16],
+    ) -> Result<SzkQueryShare<T::Field, SEED_SIZE>, SzkError> {
+        let query_rand = self.derive_query_rand(verify_key, nonce);
+
+        let is_leader = match proof_share {
+            SzkProofShare::Leader { .. } => true,
+            SzkProofShare::Helper { .. } => false,
+        };
+
+        let flp_proof_share = match proof_share {
+            SzkProofShare::Leader {
+                ref uncompressed_proof_share,
+                ..
+            } => Cow::Borrowed(uncompressed_proof_share),
+            SzkProofShare::Helper {
+                ref proof_share_seed_and_blind,
+                ..
+            } => Cow::Owned(self.derive_helper_proof_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, joint_rand_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, peer_joint_rand_part) = match proof_share {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    leader_blind_and_helper_joint_rand_part,
+                } => leader_blind_and_helper_joint_rand_part
+                    .expect("jr_len > 0 so there is joint_randomness"),
+                SzkProofShare::Helper {
+                    proof_share_seed_and_blind,
+                    leader_joint_rand_part,
+                } => (
+                    proof_share_seed_and_blind,
+                    leader_joint_rand_part.expect("jr_len > 0 so there is joint randomness"),
+                ),
+            };
+            let host_joint_rand_part = self.derive_joint_rand_part(&seed, input_share, nonce)?;
+            let (leader_joint_rand_part, helper_joint_rand_part) = if is_leader {
+                (&host_joint_rand_part, &peer_joint_rand_part)
+            } else {
+                (&peer_joint_rand_part, &host_joint_rand_part)
+            };
+            let (joint_rand_seed, joint_rand) =
+                self.derive_joint_rand(leader_joint_rand_part, helper_joint_rand_part);
+            (joint_rand, Some([joint_rand_seed, host_joint_rand_part]))
+        } else {
+            (Vec::new(), None)
+        };
+        let verifier_share = self.typ.query(
+            input_share,
+            flp_proof_share.as_ref(),
+            &query_rand,
+            &joint_rand,
+            2,
+        )?;
+        Ok(SzkQueryShare {
+            joint_rand_seed_and_part,
```

Comment:
Re: https://github.com/divviup/libprio-rs/pull/1013#discussion_r1602074882

I think this is where @divergentdave and I are confused. I think both of us are reading this code as if `SzkQueryShare` is a message that is transmitted to the peer. Is this true?

If so, do we actually need to send BOTH the joint rand seed AND our joint rand part to the peer? The way Prio3 works is that each aggregator stores the joint rand seed and broadcasts its joint rand part only. That way each aggregator can re-compute the joint rand seed from the broadcast parts and make sure it matches the stored value. The upside is that we avoid transmitting more bytes (`SEED_SIZE`) than we need to.

If you don't intend for joint rand seed to be transmitted, then it's a good idea to separate the bits that are stored and the bits that are transmitted at the API level. Concretely, you might modify the return type of this function to return the query share and the joint rand seed to be stored.

Reply:
Resolved. SzkQueryShare was not originally be meant to sent to the peer; only the verifier share and joint rand part should be set. There are now two structs: SzkQueryShare for sending and SzkQueryState for storing.